teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,298 @@
+  - **Partial** -- We can guarantee that a user can run the exact steps with the exact inputs -- but the output artifacts could change due to timestamps included in the output artifact or toolchain changes. The output artifact is functionally identical but the signature doesn't match.
+- **Hermetic builds**
+  - For this proposal, hermetic builds are out-of-scope. Hermetic build does not guarantee reproducibility (eg. If the build script links library1 or library2 depending on whether build process id or timestamp is a particular value then running the build again might not give the same result) but if a build is not hermetic then we cannot guarantee reproducibility. Tekton introduced hermetic builds in [TEP 0025](https://github.com/tektoncd/community/blob/main/teps/0025-hermekton.md). Since then, it has been an experimental feature. This feature needs to be pulled out of an experimental feature and [propagated to beta](https://github.com/tektoncd/pipeline/issues/3965https://github.com/tektoncd/pipeline/issues/3965) and beyond to unblock reproducible builds.
+- **Complete list of system architecture** like host architecture (amd64 vs arm64), SELinux, container runtimes, and kubernetes version are out-of-scope. This proposal captures tekton related system versioning.
+ 
+## Proposal
+For the completeness of build instructions, we need to be able to extract enough information from the provenance to recreate the task run or pipeline run under the same set of configuration as was used when creating the build. The list below shows which fields are required in the provenance for re-creating the task run.
+- [Task Specification](#task-specification)
+- [Step Specification](#step-specification)
+- [TaskRef](#taskref)
+- [TaskRun Specification](#taskrun-specification)
+- [Tekton System Versioning](#tekton-system-versioning)
+- [Configuration feature flags](#configuration-feature-flags)
+- [Materials](#materials)
+- [Reproducibility Reason if not reproducible](#reproducibility-reason-field-if-not-reproducible) 
+
+This will allow us to create a provenance for a task run as [shown](#provenance-for-executed-taskrun).
+
+### Task Specification
+The table below provides the name and description of the API fields that the Task requires. In addition, it also lists if that information is required, not required, or provided by the provenance. If it is required, it also shows where the information should be provided. Looking at the table below and required [step](#step-specification) information, there are only a handful of fields that are not required for reproducibility. Instead of cherry-picking most of the fields, including the entire taskSpec in the buildConfig section of the provenance is likely the best way to go forward. Additionally, if we update the spec in the future, then chains will not have to be updated correspondingly.
```

Comment:
+1 to including the entire taskSpec - trying to pick and choose individual fields will just be a maintenance nightmare. Maybe we can take the same route for Pipelines as well? (include the entire pipelineSpec)

Reply:
Yes, I agree. That was my thought as well. ie. PipelineSpec in build Config and pipeline run-time parameters in invocation.Parameters.